File: linkedlist/leetcode_Rotate_List.java
/*Rotate_List

Given a list, rotate the list to the right by k places, where k is
non-negative.
For example:
Given 1->2->3->4->5->NULL and k = 2,
return 4->5->1->2->3->NULL.
*/


class leetcode_Rotate_List {
    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        //head.next = new ListNode(2);
        //head.next.next = new ListNode(3);
        //head.next.next.next = new ListNode(4);
        //head.next.next.next.next = new ListNode(5);
        rotateRight(head, 1);
    }
    public static ListNode rotateRight(ListNode head, int n) {
        if (head == null || n == 0) {
            return head;
        }
        ListNode prev = head;
        ListNode t = head;
        int cnt = 0;
        while (t != null) {
            // Move t down to the right node, which is n-nodes away from head.
            // Or break reach the end of linked list.
            t = t.next;
            ++cnt;
            if (cnt == n) {
                break;
            }
        }
        if (t == null && cnt <= n) {
            // length of this linked list cnt is less than n.
            // rotateRigth(1->2->3, 4) is equal to rotateRight(1->2->3, 1)
            return rotateRight(head, n % cnt);
        }
        while (t.next != null) {
            // Slide down to the end of linked list.
            t = t.next;
            prev = prev.next;
        }
        ListNode newhead = prev.next;
        t.next = head;
        prev.next = null;
        return newhead;
    }
}


class ListNode {
    ListNode next;
    int val;
    ListNode(int x) {
        val = x;
        next = null;
    }
}

File: linkedlist/leetcode_Rotate_List.cpp
/**
 * Copyright 2016
 * author: madarfacar
 *
 * leetcode rotate list
 */

#include <stdio.h>

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode* next;
  ListNode(int x) : val(x), next(NULL) {}

  void printAll() {
    printf("%d -> ", val);
    ListNode* cur = next;
    while (cur) {
      printf("%d -> ", cur->val);
      cur = cur->next;
    }
    printf("NULL\n");
  }
};


class Solution {
 public:
  ListNode* rotateRight(ListNode* head, int k) {
    if (head == NULL) {
      return head;
    }
    int length = get_length(head);
    k = k % length;
    if (k == 0) {
      return head;
    }

    ListNode* start = head;
    ListNode* end = head;
    while (k > 0) {
      end = end->next;
      --k;
    }

    while (end->next != NULL) {
      start = start->next;
      end = end->next;
    }

    ListNode* rotated_head = start->next;
    start->next = NULL;
    end->next = head;

    return rotated_head;
  }

  int get_length(ListNode* head) {
    int length = 0;
    while (head != NULL) {
      head = head->next;
      ++length;
    }
    return length;
  }
};


int main() {
  ListNode head (1);
  ListNode a (2);
  head.next = &a;
  ListNode b (3);
  a.next = &b;
  ListNode c (4);
  b.next = &c;
  ListNode d (5);
  c.next = &d;
  Solution sol;

  ListNode* rotated_head = sol.rotateRight(&head, 2);
  rotated_head->printAll();
}
